### PR TITLE
Rework dma driver

### PR DESCRIPTION
**Pull request by savaga**
Needs for future max7456 driver DMA support

### CODE REVIEW

**Comment by ledvinap** [comment]
Looks good.

One thing I am not sure about is callback-user-data parameter handling. 
Personally I prefer handler object with callback function pointer as it's member. It is used in my EXTI implementation (https://github.com/ledvinap/cleanflight/blob/local/src/main/drivers/exti.c#L145 / https://github.com/ledvinap/cleanflight/blob/local/src/main/drivers/exti.h#L27) and in timer code.

Pointer to handler object is passed to callback function. It is easy to include handler structure in higher-level struct/object and use container_of macro to get 'self' pointer.
One example is here: https://github.com/cleanflight/cleanflight/blob/master/src/main/drivers/pwm_rx.c#L308, it is simple to add `overflowCb` and `edgeCb` into `pwmInputPort_t` structure.

One advantage is that you can add data members to user-allocated object, making static memory management easier (`timerOvrHandlerRec_t` is implemented as linked list of (user supplied) handlers - https://github.com/cleanflight/cleanflight/blob/master/src/main/drivers/timer.h#L55)

I am a bit biased in this matter, so take it only as suggestion/idea. Maybe other CF developers may comment too ...

Another minor optimization is to use dummy function instead of NULL for unused handler - that way few instructions are saved in IRQ ..

**Comment by savaga** [comment]
@ledvinap, thanks for your review! I actually like the implementation of the linked list of user handlers. And I'm gonna use it in DMA too. The reason I need multiple handlers is supporting multiple devices on the same bus (like spi or i2c). Will update the PR soon

**Comment by ledvinap** [comment]
@savaga: I would expect single SPI/I2C driver per interface, with this driver handling all interface-specifin details (interrupts/dma/bus arbitration) and providing functions to multiple device drivers. Allowing direct DMA access to device driver will probably result in poor layer isolation.

The problem with I2C is IMO more about bus arbitration and glue that will connect interrupt and main loop. 

Ideal solution is using RTOS, but CF is moving only slowly in this direction.

One possibility is to move (considerable) part of device driver into interrupt and let it communicate with task code somehow. But this will be quite hard to debug (multiple devices will be mixed together) and may increase interrupt latency considerable.

Or some command processor may in implemented in I2C driver - device driver will queue commands to be executed (read/write register(s); maybe same quick callback to handle conditional code effectively) and wait for competition (re-schedule when commands are finished).

**Comment by jflyper** [comment]
> Or some command processor may in implemented in I2C driver - device driver will queue commands to be executed (read/write register(s); maybe same quick callback to handle conditional code effectively) and wait for competition (re-schedule when commands are finished).

+1
I was thinking of a similar approach, but I wasn't sure how to transfer from interrupt context to task context. Scheduler should be modified to honor tasks waiting for events?

**Comment by ledvinap** [comment]
@jflyper : Scheduler is pretty overengineered. The whole soft priority thing it far too complicated and is not used (ARM is refused when load is high - simple priority based scheduler is almost identical when load is light)

Normally you will simply move task to run-queue and use same mechanism to pass data.

IMO it would be best to take scheduler from same tested RTOS - freeRTOS / ChibiOS etc .. Just the part to select proces for running, no preemption.

**Comment by hydra** [comment]
@savaga Hiya!  Yes, I agree with @ledvinap regarding the use of callback object, likely there is other areas we can cleanup in Cleanflight regarding that but we can deal with that after this - this is time sensitive so please focus on this PR if possible.

When do you think you'll have a chance to finish this PR?

**Comment by savaga** [comment]
@hydra, I'll update the PR tonight (PDT)

File: src/main/drivers/dma.c [hunk 1]
```diff
@@ -24,60 +24,99 @@
...
+DEFINE_DMA_IRQ_HANDLER(2, 4, DMA2_CH4_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(2, 5, DMA2_CH5_HANDLER);
+#endif
 
-void DMA1_Channel7_IRQHandler(void)
+
+void dmaInit(void)
 {
-    dmaHandlers.dma1Channel7IRQHandler(DMA1_Channel7);
+    // TODO: Do we need this?
```
Comment (ledvinap) [inline src/main/drivers/dma.c]: In future, DMA resource tracing may be implemented. And some other initialization may appear.
And only constant part of handler descriptors may be stored in flash, initializing rest of variables here in loop

**Comment by hydra** [comment]
@savaga Thanks for the update, @ledvinap thanks for the review!

**Comment by hydra** [comment]
@savaga can you fix the failing unit tests.  Ready to merge otherwise?

**Comment by savaga** [comment]
The test passes in my local branch, but it fails in master. Last failure isn't related to my PR

**Comment by hydra** [comment]
@savaga coolthanks, i'll check it out and merge asap.

**Comment by hydra** [comment]
ok @savaga merged.

**Comment by hydra** [comment]
@savaga See https://github.com/cleanflight/cleanflight/issues/2210#issuecomment-233285322

The max7456 via DMA should follow this.

**Comment by ledvinap** [comment]
@hydra: This PR is was failing build.
https://github.com/cleanflight/cleanflight/blob/master/src/main/target/RCEXPLORERF3/target.h#L178 is not converted to new DMA handler

**Comment by hydra** [comment]
@savaga i have referenced your DMA work and got it working on the boards here.  No need for a PR.

**Pull request merged**